#Segmenting and Clustering Neighbourhoods in Toronto




##Part 1

Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

Get the HTML page of Wiki, and convert the html data into list of Data Frame objects.


In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
data = requests.get(url).text

In [3]:
soup = BeautifulSoup(data, "html5lib")

Creating a empty list, dictionary and extract it all the dataframe df

In [4]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})


In [5]:
df

PostalCode  ...                                       Neighborhood
0          M3A  ...                                          Parkwoods
1          M4A  ...                                   Victoria Village
2          M5A  ...                          Regent Park, Harbourfront
3          M6A  ...                   Lawrence Manor, Lawrence Heights
4          M7A  ...                      Ontario Provincial Government
..         ...  ...                                                ...
98         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99         M4Y  ...                               Church and Wellesley
100        M7Y  ...                                     Enclave of M4L
101        M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102        M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

The Number of Rows and Columns in the data frame

In [6]:
df.shape

(103, 3)

## Part 2

Installing the geocoder

In [7]:
pip install geocoder

     |████████████████████████████████| 102kB 3.6MB/s 


In [8]:
import geocoder

Tried the Below approach for minutes, it gave no result.

In [ ]:
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'
# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

KeyboardInterrupt: ignored

In [9]:
data = pd.read_csv("/content/Geospatial_Coordinates.csv")
data

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

Examining the Shape of the dataframe

In [10]:
print('The Shape of our wiki data is: ', df.shape)
print('The Shape of our csv data is : ', data.shape)

The Shape of our wiki data is:  (103, 3)
The Shape of our csv data is :  (103, 3)


The Dimensions of df and data dataframes are same.

In [11]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [12]:
data.dtypes

PostalCode     object
Latitude      float64
Longitude     float64
dtype: object

Joining the dataframes on PostalCode Column

In [13]:
combined_data = df.join(data.set_index('PostalCode'), on = 'PostalCode', how = 'inner')
combined_data

PostalCode                Borough  ...   Latitude  Longitude
0          M3A             North York  ...  43.753259 -79.329656
1          M4A             North York  ...  43.725882 -79.315572
2          M5A       Downtown Toronto  ...  43.654260 -79.360636
3          M6A             North York  ...  43.718518 -79.464763
4          M7A           Queen's Park  ...  43.662301 -79.389494
..         ...                    ...  ...        ...        ...
98         M8X              Etobicoke  ...  43.653654 -79.506944
99         M4Y       Downtown Toronto  ...  43.665860 -79.383160
100        M7Y  East Toronto Business  ...  43.662744 -79.321558
101        M8Y              Etobicoke  ...  43.636258 -79.498509
102        M8Z              Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

##PART 3


Previously we cluster the neighborhood of NYC. We cluster Toronto based on
similarites of venues categories using K means

In [14]:
import geocoder
from geopy.geocoders import Nominatim

In [15]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [16]:
import folium

In [18]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [19]:
CLIENT_ID = 'FSJXMK50GG0BFPDWKOG1JYQVBPEOTPESHELJL3DHMZXZPXI2' 
CLIENT_SECRET = '5ERS3ZI430U31ROETTEQCNRLV3SHKP3CDPLBJ1AWUAOGCXMN'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FSJXMK50GG0BFPDWKOG1JYQVBPEOTPESHELJL3DHMZXZPXI2
CLIENT_SECRET:5ERS3ZI430U31ROETTEQCNRLV3SHKP3CDPLBJ1AWUAOGCXMN


We create a function like the one we did when clustering the New York neighborhood

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [22]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [23]:
venues_in_toronto.shape


(1275, 5)

So we got 1317 records and 5 columns

In [24]:
venues_in_toronto.head()


Neighbourhood  ...        Venue Category
0         Parkwoods  ...  Fast Food Restaurant
1         Parkwoods  ...                  Park
2         Parkwoods  ...              Bus Stop
3         Parkwoods  ...     Food & Drink Shop
4  Victoria Village  ...          Hockey Arena

[5 rows x 5 columns]

Checking the venues based on Neighbourhood

In [27]:
venues_in_toronto.groupby('Neighbourhood').head()


Neighbourhood  ...        Venue Category
0                                             Parkwoods  ...  Fast Food Restaurant
1                                             Parkwoods  ...                  Park
2                                             Parkwoods  ...              Bus Stop
3                                             Parkwoods  ...     Food & Drink Shop
4                                      Victoria Village  ...          Hockey Arena
...                                                 ...  ...                   ...
1262  Mimico NW, The Queensway West, South of Bloor,...  ...          Burger Joint
1263  Mimico NW, The Queensway West, South of Bloor,...  ...           Wings Joint
1264  Mimico NW, The Queensway West, South of Bloor,...  ...       Supplement Shop
1265  Mimico NW, The Queensway West, South of Bloor,...  ...        Discount Store
1266  Mimico NW, The Queensway West, South of Bloor,...  ...        Sandwich Place

[430 rows x 5 columns]

In [26]:
venues_in_toronto.groupby('Venue Category').max()


Neighbourhood  ...                            Venue
Venue Category                                                         ...                                 
Accessories Store                    Lawrence Manor, Lawrence Heights  ...                           Ardene
Adult Boutique                                   Church and Wellesley  ...                        Seduction
Airport                                                Downsview East  ...  Toronto Downsview Airport (YZD)
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...  ...                Billy Bishop Café
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...  ...                           Gate 8
...                                                               ...  ...                              ...
Wine Bar                                     Little Portugal, Trinity  ...                The National Club
Wine Shop                                        Church and Wellesley  ...                        Wine Rack
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...  ...                       Wingporium
Women's Store                        Lawrence Manor, Lawrence Heights  ...                    Maximum Woman
Yoga Studio                              The Danforth West, Riverdale  ...                        YogaSpace

[232 rows x 4 columns]

So, we have 232 types of Venue Categories.

####One Hot encoding the venue Categories


In [28]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat


Accessories Store  Adult Boutique  ...  Women's Store  Yoga Studio
0                     0               0  ...              0            0
1                     0               0  ...              0            0
2                     0               0  ...              0            0
3                     0               0  ...              0            0
4                     0               0  ...              0            0
...                 ...             ...  ...            ...          ...
1270                  0               0  ...              0            0
1271                  0               0  ...              0            0
1272                  0               0  ...              0            0
1273                  0               0  ...              0            0
1274                  0               0  ...              0            0

[1275 rows x 232 columns]

In [29]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  ...  Women's Store  Yoga Studio
0         Parkwoods                  0  ...              0            0
1         Parkwoods                  0  ...              0            0
2         Parkwoods                  0  ...              0            0
3         Parkwoods                  0  ...              0            0
4  Victoria Village                  0  ...              0            0

[5 rows x 233 columns]

In [30]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  ...  Yoga Studio
0                                        Agincourt  ...          0.0
1                           Alderwood, Long Branch  ...          0.0
2  Bathurst Manor, Wilson Heights, Downsview North  ...          0.0
3                                  Bayview Village  ...          0.0
4                Bedford Park, Lawrence Manor East  ...          0.0

[5 rows x 233 columns]

Let's make a function to get the the top most common venue

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [32]:
import numpy as np


In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood  ...       10th Most Common Venue
0                                        Agincourt  ...                    Drugstore
1                           Alderwood, Long Branch  ...         Ethiopian Restaurant
2  Bathurst Manor, Wilson Heights, Downsview North  ...            Convenience Store
3                                  Bayview Village  ...  Eastern European Restaurant
4                Bedford Park, Lawrence Manor East  ...              Thai Restaurant

[5 rows x 11 columns]

Let us make the model for clustering

In [34]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [35]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans


KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Let's add the clustering Label column to the top 10 common venue categories



In [36]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)


Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [40]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head()

PostalCode           Borough  ...  9th Most Common Venue  10th Most Common Venue
0        M3A        North York  ...              Drugstore              Donut Shop
1        M4A        North York  ...             Donut Shop                 Dog Run
2        M5A  Downtown Toronto  ...                    Spa          Farmers Market
3        M6A        North York  ...  Vietnamese Restaurant                   Diner
4        M7A      Queen's Park  ...            Music Venue      Mexican Restaurant

[5 rows x 16 columns]

In [41]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])


Plotting the clusters on the map



In [42]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [44]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [45]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
50       North York  ...                 Dog Run
83  Central Toronto  ...                 Dog Run

[2 rows x 12 columns]

Cluster 2

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  ...       10th Most Common Venue
0                North York  ...                   Donut Shop
3                North York  ...                        Diner
6               Scarborough  ...                      Dog Run
7                North York  ...          Distribution Center
8                 East York  ...                    Gastropub
9          Downtown Toronto  ...                     Tea Room
10               North York  ...                   Donut Shop
12              Scarborough  ...                    Drugstore
13               North York  ...               Sandwich Place
14                East York  ...                      Dog Run
15         Downtown Toronto  ...                 Cocktail Bar
16                     York  ...                   Donut Shop
17                Etobicoke  ...           Dim Sum Restaurant
18              Scarborough  ...           Dim Sum Restaurant
19             East Toronto  ...                   Donut Shop
20         Downtown Toronto  ...                    Jazz Club
25         Downtown Toronto  ...  Eastern European Restaurant
26              Scarborough  ...            Electronics Store
27               North York  ...                    Drugstore
29                East York  ...               Discount Store
30         Downtown Toronto  ...                 Neighborhood
31             West Toronto  ...                  Gas Station
32              Scarborough  ...                      Dog Run
36         Downtown Toronto  ...                     Beer Bar
37             West Toronto  ...          Japanese Restaurant
39               North York  ...  Eastern European Restaurant
40               North York  ...                    Drugstore
41             East Toronto  ...                         Café
42         Downtown Toronto  ...                          Pub
43             West Toronto  ...                   Restaurant
44              Scarborough  ...            Electronics Store
46               North York  ...  Eastern European Restaurant
47             East Toronto  ...                   Restaurant
48         Downtown Toronto  ...             Sushi Restaurant
49               North York  ...  Eastern European Restaurant
53               North York  ...                   Donut Shop
54             East Toronto  ...                          Bar
57               North York  ...                    Drugstore
58              Scarborough  ...                    Drugstore
60               North York  ...            Electronics Store
62          Central Toronto  ...                   Donut Shop
63                     York  ...  Eastern European Restaurant
65              Scarborough  ...                      Dog Run
67          Central Toronto  ...                  Pizza Place
69             West Toronto  ...                       Bakery
73          Central Toronto  ...           Italian Restaurant
78              Scarborough  ...                    Drugstore
80         Downtown Toronto  ...        College Arts Building
84         Downtown Toronto  ...           Mexican Restaurant
85              Scarborough  ...                    Drugstore
87         Downtown Toronto  ...                 Airport Gate
92   Downtown Toronto Stn A  ...                         Café
94      Etobicoke Northwest  ...                   Donut Shop
96         Downtown Toronto  ...                         Bank
97         Downtown Toronto  ...                          Pub
98                Etobicoke  ...                   Donut Shop
99         Downtown Toronto  ...            Indian Restaurant
100   East Toronto Business  ...               Farmers Market
101               Etobicoke  ...  Eastern European Restaurant
102               Etobicoke  ...                  Social Club

[60 rows x 12 columns]

Cluster 3

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...       10th Most Common Venue
21                    York  ...                      Dog Run
35  East York/East Toronto  ...                    Drugstore
52              North York  ...                      Dog Run
61         Central Toronto  ...                    Drugstore
64                    York  ...                   Donut Shop
66              North York  ...                    Drugstore
68         Central Toronto  ...  Eastern European Restaurant
77               Etobicoke  ...                   Donut Shop
91        Downtown Toronto  ...                   Donut Shop

[9 rows x 12 columns]

Cluster 4

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
11  Etobicoke             3.0  ...            Donut Shop                Dog Run

[1 rows x 12 columns]

Cluster 5

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  ...       10th Most Common Venue
1         North York  ...                      Dog Run
2   Downtown Toronto  ...               Farmers Market
4       Queen's Park  ...           Mexican Restaurant
22       Scarborough  ...                   Donut Shop
23         East York  ...               Sandwich Place
24  Downtown Toronto  ...           Italian Restaurant
28        North York  ...            Convenience Store
33        North York  ...          Sporting Goods Shop
34        North York  ...  Eastern European Restaurant
38       Scarborough  ...                   Donut Shop
51       Scarborough  ...                   Donut Shop
55        North York  ...              Thai Restaurant
56              York  ...                   Donut Shop
59        North York  ...                Shopping Mall
70         Etobicoke  ...                    Drugstore
71       Scarborough  ...                   Donut Shop
72        North York  ...                Deli / Bodega
74   Central Toronto  ...            Indian Restaurant
75      West Toronto  ...               Breakfast Spot
76       Mississauga  ...                  Gas Station
79   Central Toronto  ...           Italian Restaurant
81      West Toronto  ...                         Park
82       Scarborough  ...                     Pharmacy
86   Central Toronto  ...   Construction & Landscaping
88         Etobicoke  ...                          Gym
89         Etobicoke  ...            Electronics Store
90       Scarborough  ...                  Coffee Shop
93         Etobicoke  ...         Ethiopian Restaurant

[28 rows x 12 columns]